File: simulation/gpu_solver/src/flux.rs
use crate::types::Grid2D;

/// Compute horizontal and vertical diffusive fluxes on cell faces.
///
/// This is the Rust equivalent of the Python `compute_flux` function.
/// Arrays are flattened in row-major order.
pub fn compute_flux(
    grid: Grid2D,
    concentration: &[f32],
    diffusivity: &[f32],
) -> (Vec<f32>, Vec<f32>) {
    let width = grid.width;
    let height = grid.height;

    assert_eq!(concentration.len(), grid.len());
    assert_eq!(diffusivity.len(), grid.len());

    let mut jx = vec![0.0; height * (width - 1)];
    let mut jy = vec![0.0; (height - 1) * width];

    // Horizontal fluxes between (row, col) and (row, col + 1).
    for row in 0..height {
        for col in 0..(width - 1) {
            let left = grid.idx(row, col);
            let right = grid.idx(row, col + 1);
            let face = row * (width - 1) + col;

            let d_avg = 0.5 * (diffusivity[left] + diffusivity[right]);
            let dc_dx = (concentration[right] - concentration[left]) / grid.dx;

            jx[face] = -d_avg * dc_dx;
        }
    }

    // Vertical fluxes between (row, col) and (row + 1, col).
    for row in 0..(height - 1) {
        for col in 0..width {
            let top = grid.idx(row, col);
            let bottom = grid.idx(row + 1, col);
            let face = row * width + col;

            let d_avg = 0.5 * (diffusivity[top] + diffusivity[bottom]);
            let dc_dy = (concentration[bottom] - concentration[top]) / grid.dy;

            jy[face] = -d_avg * dc_dy;
        }
    }

    (jx, jy)
}

/// Convert face fluxes into per-cell concentration change from diffusion.
///
/// This is the Rust equivalent of the Python `compute_flux_divergence` function.
pub fn compute_flux_divergence(grid: Grid2D, jx: &[f32], jy: &[f32]) -> Vec<f32> {
    let width = grid.width;
    let height = grid.height;

    assert_eq!(jx.len(), height * (width - 1));
    assert_eq!(jy.len(), (height - 1) * width);

    let mut dcdt = vec![0.0; grid.len()];

    // Horizontal flux contribution.
    for row in 0..height {
        for col in 0..(width - 1) {
            let flux = jx[row * (width - 1) + col];
            let left = grid.idx(row, col);
            let right = grid.idx(row, col + 1);

            dcdt[left] -= flux / grid.dx;
            dcdt[right] += flux / grid.dx;
        }
    }

    // Vertical flux contribution.
    for row in 0..(height - 1) {
        for col in 0..width {
            let flux = jy[row * width + col];
            let top = grid.idx(row, col);
            let bottom = grid.idx(row + 1, col);

            dcdt[top] -= flux / grid.dy;
            dcdt[bottom] += flux / grid.dy;
        }
    }

    dcdt
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn uniform_concentration_has_zero_flux() {
        let grid = Grid2D::new(5, 5, 1.0, 1.0);
        let concentration = vec![1.0; grid.len()];
        let diffusivity = vec![1.0; grid.len()];

        let (jx, jy) = compute_flux(grid, &concentration, &diffusivity);

        assert!(jx.iter().all(|value| value.abs() < 1e-6));
        assert!(jy.iter().all(|value| value.abs() < 1e-6));
    }
}

File: simulation/solver_reference/flux.py
import numpy as np


def compute_flux(C, D, dx, dy):
	"""
		Computes the diffusive flux between neighboring grid cells using Fick's Law and a 
		face-centered finite-volume approach

		Parameters:
			C	-> a 2D array of solute concentration values
			D	-> a 2D array of effective diffusivity values for the solute at given grid cell
			dx	-> grid spacing in the x direction
			dy	-> grid spacing in the y direction

		Returns:
			Jx	-> flux between left/right neighboring cells
			Jy	-> flux between top/bottom neighboring cells

		Notes:
			Concentration values are stored at cell centers while flux values are computed at the
			faces between neighboring cells. Flux direction and magnitude are determined by the
			concentration gradient and local effective diffusivity.

			For neighboring cells C[i,j] and C[i,j+1]:

				dC/dx = (C[i,j+1] - C[i,j]) / dx
				Jx = -D * dC/dx

			Positive flux corresponds to transport in the positive coordinate direction, while 
			negative flux indicates transport in the opposite direction.
	"""
	# Face-centered diffusivity values between neighboring cells.
	D_x = 0.5 * (D[:, :-1] + D[:, 1:])
	D_y = 0.5 * (D[:-1, :] + D[1:, :])

	# Concentration gradients across cell faces.
	dCdx = (C[:, 1:] - C[:, :-1]) / dx
	dCdy = (C[1:, :] - C[:-1, :]) / dy

	# Fick's law: J = -D grad(C)
	Jx = -D_x * dCdx
	Jy = -D_y * dCdy

	return Jx, Jy



def compute_flux_divergence(Jx, Jy, dx, dy, shape):
	"""
		Computes dC/dt for each grid cell from the fluxes between neighboring grid cells

		Parameters:
			Jx		-> flux between neighboring grid cells along the x axis
			Jy		-> flux between neighboring grid cells along the y axis
			dx		-> grid spacing in the x direction
			dy		-> grid spacing in the y direction
			shape	-> shape of the concetration gridß
	
		Returns:
			dCdt	-> an array representing the change in concentration of the solute in each
					grid cell
	
		Notes:
			Flux values are stored at the faces between neighboring grid cells while 
			concentration changes (dC/dt) are accumulated at grid cell centers.

			This function computes the net concentration change within each grid cell by
			enforcing conservation of mass: solute leaving one cell through a shared boundary
			enters the neighboring cell through the same boundary.

			Rather than explicitly computing:

				dC/dt = -(dJx/dx + dJy/dy)

			concentration changes are obtained by directly distributing flux across shared
			grid cell faces. This approach is mathematically equivalent to computing the
			divergence of flux while ensuring local conservation of mass.

			Positive outgoing flux decreases concentration in the source grid cell, while the
			same flux increases concentration in the neighboring receiving grid cell.
	"""
	dCdt = np.zeros(shape, dtype=Jx.dtype)

	# Horizontal flux contribution.
	# Jx[:, j] is the flux from cell j to cell j + 1.
	dCdt[:, :-1] -= Jx / dx
	dCdt[:, 1:] += Jx / dx

	# Vertical flux contribution.
	# Jy[i, :] is the flux from row i to row i + 1.
	dCdt[:-1, :] -= Jy / dy
	dCdt[1:, :] += Jy / dy

	return dCdt